### ISSUE TITLE
Move in, Move in overwrite and Move out fail using WebDAV with locking

### ISSUE THREAD

**Issue by nickvergessen** [owncloud/core#17029]
https://github.com/owncloud/core/pull/16963#issuecomment-113206603

Test | Result
---- | ------
test_sharePermissions -t=0 (ALL) | :x: move_in, move_in_overwrite and move_out fail
test_sharePermissions -t=1 (READ) | :white_check_mark: 
test_sharePermissions -t=2 (READ + CREATE) | :x: move_in
test_sharePermissions -t=3 (READ + UDPATE) | :white_check_mark: 
test_sharePermissions -t=4 (READ + DELETE) | :x: move_out
test_sharePermissions -t=5 (READ + CREATE + UPDATE) | :x: move_in
test_sharePermissions -t=6 (READ + CREATE + DELETE) | :x: move_in, move_in_overwrite and move_out
test_sharePermissions -t=7 (READ + UPDATE + DELETE) | :x: move_out

Note: move_in_overwrite deletes + creates thats why READ+UPDATE is okay.

@icewind1991 @PVince81 @DeepDiver1975

**Comment by PVince81** [comment]
This issue title sounds like a cheat code

**Comment by nickvergessen** [comment]
```
{"reqId":"5at1fJdV3+W5ruk+uhYL","remoteAddr":"192.168.178.32","app":"PHP","message":"Call to a member function getPath() on a non-object at \/home\/nickv\/ownCloud\/Smashbox\/master\/lib\/private\/share\/share.php#134","level":3,"time":"2015-06-19T07:02:48+00:00"}
```

@icewind1991 `$meta = $view->getFileInfo($path);` in line 133 returns false, so of course the method does not exist then.

**Comment by icewind1991** [comment]
I can't reproduce any of the failures locally...

**Comment by DeepDiver1975** [comment]
> I can't reproduce any of the failures locally...

are you executing the smashbox test test_sharePermissions? - It also fails on jenkins: https://ci.owncloud.org/job/smashbox/TEST_NAME=test_sharePermissions,compiler=g++,label=master,mirallBranch=v1.8.1,qt=qt4,serverVersion=owncloud8.1-locking/55/console

**Comment by PVince81** [comment]
Same for me, running this:
```
% bin/smash lib/oc-tests/test_sharePermissions.py
```

no errors.

Oh, let me try again with `-a` to run all test cases...

**Comment by DeepDiver1975** [comment]
> Oh, let me try again with -a to run all test cases...

haha - that bloody switch

**Comment by PVince81** [comment]
Tried again twice with the bloody switch, but still no failure

**Comment by PVince81** [comment]
@nickvergessen you could also reproduce this locally, consistently ?
Can you do a "flushall" on redis before you do ?

Maybe there are stray keys

**Comment by DeepDiver1975** [comment]
> Maybe there are stray keys

even if - on jenkins it's falling apart with http 500

````
INFO:requests.packages.urllib3.connectionpool:Starting new HTTPS connection (1): test1.owncloud.com
2015-06-19 04:56:52,239 - INFO - recipient_worker - Checking that /var/lib/jenkins/build/smashdir/test_sharePermissions/recipient_worker/shared-dir is present in local directory for recipient_worker
2015-06-19 04:56:52,239 - INFO - recipient_worker - createfile /var/lib/jenkins/build/smashdir/test_sharePermissions/recipient_worker/local_test_file_0.dat character='0' count=1000 bs=1
2015-06-19 04:56:52,240 - INFO - recipient_worker - Upload file to "shared-dir/test_upload.dat"
2015-06-19 04:56:55,477 - INFO - recipient_worker - createfile /var/lib/jenkins/build/smashdir/test_sharePermissions/recipient_worker/local_test_file_1.dat character='0' count=1000 bs=1
2015-06-19 04:56:55,478 - INFO - recipient_worker - Upload file to "shared-dir/overwrite_this.dat"
2015-06-19 04:56:59,301 - INFO - recipient_worker - Move "shared-dir/rename_this.dat" to "shared-dir/rename_this_renamed.dat"
2015-06-19 04:57:00,539 - INFO - recipient_worker - createfile /var/lib/jenkins/build/smashdir/test_sharePermissions/recipient_worker/local_test_file_2.dat character='0' count=1000 bs=1
2015-06-19 04:57:01,734 - INFO - recipient_worker - Move "test_move_in.dat" to "shared-dir/test_move_in.dat"
2015-06-19 04:57:02,380 - ERROR - recipient_worker - Unexpected response: Status code: 500 - 
2015-06-19 04:57:02,380 - INFO - recipient_worker - Full Response: 
2015-06-19 04:57:02,427 - ERROR - recipient_worker - Operation "move_in" must be allowed when share permissions are 31 'Operation "%s" must be %s when share permissions are %i' % (operation, success_message, perms) failed in recipient_worker() ["/var/lib/jenkins/jobs/smashbox/workspace/TEST_NAME/test_sharePermissions/compiler/g++/label/master/mirallBranch/v1.8.1/qt/qt4/serverVersion/owncloud8.1-locking/lib/oc-tests/test_sharePermissions.py" at line 251]
2015-06-19 04:57:02,427 - INFO - recipient_worker - createfile /var/lib/jenkins/build/smashdir/test_sharePermissions/recipient_worker/local_test_file_3.dat character='0' count=1000 bs=1
2015-06-19 04:57:04,390 - INFO - recipient_worker - Move "overwrite_this_through_move_in.dat" to "shared-dir/overwrite_this_through_move_in.dat"
2015-06-19 04:57:13,720 - ERROR - recipient_worker - Unexpected response: Status code: 500 - 
2015-06-19 04:57:13,721 - INFO - recipient_worker - Full Response: 
2015-06-19 04:57:13,722 - ERROR - recipient_worker - Operation "move_in_overwrite" must be allowed when share permissions are 31 'Operation "%s" must be %s when share permissions are %i' % (operation, success_message, perms) failed in recipient_worker() ["/var/lib/jenkins/jobs/smashbox/workspace/TEST_NAME/test_sharePermissions/compiler/g++/label/master/mirallBranch/v1.8.1/qt/qt4/serverVersion/owncloud8.1-locking/lib/oc-tests/test_sharePermissions.py" at line 251]
2015-06-19 04:57:13,722 - INFO - recipient_worker - Move "shared-dir/move_this_to_subdir.dat" to "shared-dir/subdir/moved_this_to_subdir.dat"
2015-06-19 04:57:16,540 - INFO - recipient_worker - Move "shared-dir/move_this_to_subdir_for_overwrite.dat" to "shared-dir/subdir/overwrite_this.dat"
2015-06-19 04:57:22,002 - INFO - recipient_worker - Move "shared-dir/move_this_out.dat" to "this_was_moved_out.dat"
2015-06-19 04:57:22,316 - ERROR - recipient_worker - Unexpected response: Status code: 500 - 
2015-06-19 04:57:22,317 - INFO - recipient_worker - Full Response: 
2015-06-19 04:57:22,318 - ERROR - recipient_worker - Operation "move_out" must be allowed when share permissions are 31 'Operation "%s" must be %s when share permissions are %i' % (operation, success_message, perms) failed in recipient_worker() ["/var/lib/jenkins/jobs/smashbox/workspace/TEST_NAME/test_sharePermissions/compiler/g++/label/master/mirallBranch/v1.8.1/qt/qt4/serverVersion/owncloud8.1-locking/lib/oc-tests/test_sharePermissions.py" at line 251]
2015-06-19 04:57:22,319 - INFO - recipient_worker - Move "shared-dir/subdir/move_this_out.dat" to "shared-dir/this_was_moved_out_of_subdir.dat"
2015-06-19 04:57:26,627 - INFO - recipient_worker - Delete "shared-dir/delete_this.dat"
2015-06-19 04:57:31,125 - INFO - recipient_worker - Create folder "shared-dir/test_create_dir"
2015-06-19 04:57:31,959 - INFO - recipient_worker - Delete "shared-dir/delete_this_dir"
2015-06-19 04:57:36,373 - INFO - recipient_worker - entering new step

**Comment by PVince81** [comment]
Then we might need to setup a VM with the same setup.

My current setup where tests pass is:
- openSUSE Factory
- php5-5.6.9-1.1.x86_64
- apache2-2.4.12-7.1.x86_64
- redis-3.0.2-1.1.x86_64

**Comment by DeepDiver1975** [comment]
test1 is using apcu - related?

**Comment by nickvergessen** [comment]
I can reproduce this 100% of the time, also the return false thing is obvious from looking at the code path and is not dealt with correctly (not checked).

Make sure you use the latest master on all involved tools (oc, sb, pyoc)

Setup
* Linux Mint
* php5: 5.5.9 and 5.5.26
* apache2: dont know
* redis: 2.2.7

**Comment by DeepDiver1975** [comment]
I'm currently rerunning the test on jenkins - tailing the server log to see some more .....

**Comment by DeepDiver1975** [comment]
> I'm currently rerunning the test on jenkins - tailing the server log to see some more .....

````
{"reqId":"buGcKoD8Zg1O4jfH0RlU","remoteAddr":"148.251.209.108","app":"PHP","message":"Call to a member function getPath() on boolean at \/srv\/www\/owncloud8.1-locking\/lib\/private\/share\/share.php#132","level":3,"time":"2015-06-19T12:51:49+00:00"}
{"reqId":"pCcs0ex0O47zHROs9rWr","remoteAddr":"148.251.209.108","app":"no app in context","message":"Can not decrypt this file, probably this is a shared file. Please ask the file owner to reshare the file with you.","level":3,"time":"2015-06-19T12:51:52+00:00"}
{"reqId":"pCcs0ex0O47zHROs9rWr","remoteAddr":"148.251.209.108","app":"webdav","message":"Exception: {\"Message\":\"Can not decrypt this file, probably this is a shared file. Please ask the file owner to reshare the file with you.\",\"Code\":0,\"Trace\":\"#0 \\\/srv\\\/www\\\/owncloud8.1-locking\\\/lib\\\/private\\\/files\\\/stream\\\/encryption.php(441): OCA\\\\Encryption\\\\Crypto\\\\Encryption->decrypt('JvypjpVMYqmrjq8...')\\n#1 \\\/srv\\\/www\\\/owncloud8.1-locking\\\/lib\\\/private\\\/files\\\/stream\\\/encryption.php(280): OC\\\\Files\\\\Stream\\\\Encryption->readCache()\\n#2 [internal function]: OC\\\\Files\\\\Stream\\\\Encryption->stream_read(8192)\\n#3 \\\/srv\\\/www\\\/owncloud8.1-locking\\\/lib\\\/private\\\/helper.php(616): fread(Resource id #350, 8192)\\n#4 \\\/srv\\\/www\\\/owncloud8.1-locking\\\/lib\\\/private\\\/files\\\/storage\\\/wrapper\\\/encryption.php(502): OC_Helper::streamCopy(Resource id #350, Resource id #354)\\n#5 \\\/srv\\\/www\\\/owncloud8.1-locking\\\/lib\\\/private\\\/files\\\/storage\\\/wrapper\\\/encryption.php(435): OC\\\\Files\\\\Storage\\\\Wrapper\\\\Encryption->copyBetweenStorage(Object(OCA\\\\Files_Trashbin\\\\Storage), 'overwrite_this_...', 'files_trashbin\\\/...', true, true)\\n#6 \\\/srv\\\/www\\\/owncloud8.1-locking\\\/lib\\\/private\\\/files\\\/storage\\\/wrapper\\\/wrapper.php(535): OC\\\\Files\\\\Storage\\\\Wrapper\\\\Encryption->moveFromStorage(Object(OCA\\\\Files_Trashbin\\\\Storage), 'overwrite_this_...', 'files_trashbin\\\/...')\\n#7 \\\/srv\\\/www\\\/owncloud8.1-locking\\\/apps\\\/files_trashbin\\\/lib\\\/trashbin.php(213): OC\\\\Files\\\\Storage\\\\Wrapper\\\\Wrapper->moveFromStorage(Object(OCA\\\\Files_Trashbin\\\\Storage), 'overwrite_this_...', 'files_trashbin\\\/...')\\n#8 \\\/srv\\\/www\\\/owncloud8.1-locking\\\/apps\\\/files_trashbin\\\/lib\\\/storage.php(125): OCA\\\\Files_Trashbin\\\\Trashbin::move2trash('shared-dir\\\/over...')\\n#9 \\\/srv\\\/www\\\/owncloud8.1-locking\\\/apps\\\/files_trashbin\\\/lib\\\/storage.php(89): OCA\\\\Files_Trashbin\\\\Storage->doDelete('overwrite_this_...', 'unlink')\\n#10 \\\/srv\\\/www\\\/owncloud8.1-locking\\\/lib\\\/private\\\/files\\\/view.php(1006): OCA\\\\Files_Trashbin\\\\Storage->unlink('overwrite_this_...')\\n#11 \\\/srv\\\/www\\\/owncloud8.1-locking\\\/lib\\\/private\\\/files\\\/view.php(596): OC\\\\Files\\\\View->basicOperation('unlink', '\\\/shared-dir\\\/ove...', Array)\\n#12 \\\/srv\\\/www\\\/owncloud8.1-locking\\\/lib\\\/private\\\/connector\\\/sabre\\\/file.php(294): OC\\\\Files\\\\View->unlink('\\\/shared-dir\\\/ove...')\\n#13 \\\/srv\\\/www\\\/owncloud8.1-locking\\\/3rdparty\\\/sabre\\\/dav\\\/lib\\\/DAV\\\/Tree.php(179): OC\\\\Connector\\\\Sabre\\\\File->delete()\\n#14 \\\/srv\\\/www\\\/owncloud8.1-locking\\\/3rdparty\\\/sabre\\\/dav\\\/lib\\\/DAV\\\/CorePlugin.php(622): Sabre\\\\DAV\\\\Tree->delete('shared-dir\\\/over...')\\n#15 [internal function]: Sabre\\\\DAV\\\\CorePlugin->httpMove(Object(Sabre\\\\HTTP\\\\Request), Object(Sabre\\\\HTTP\\\\Response))\\n#16 \\\/srv\\\/www\\\/owncloud8.1-locking\\\/3rdparty\\\/sabre\\\/event\\\/lib\\\/EventEmitterTrait.php(105): call_user_func_array(Array, Array)\\n#17 \\\/srv\\\/www\\\/owncloud8.1-locking\\\/3rdparty\\\/sabre\\\/dav\\\/lib\\\/DAV\\\/Server.php(469): Sabre\\\\Event\\\\EventEmitter->emit('method:MOVE', Array)\\n#18 \\\/srv\\\/www\\\/owncloud8.1-locking\\\/3rdparty\\\/sabre\\\/dav\\\/lib\\\/DAV\\\/Server.php(254): Sabre\\\\DAV\\\\Server->invokeMethod(Object(Sabre\\\\HTTP\\\\Request), Object(Sabre\\\\HTTP\\\\Response))\\n#19 \\\/srv\\\/www\\\/owncloud8.1-locking\\\/apps\\\/files\\\/appinfo\\\/remote.php(83): Sabre\\\\DAV\\\\Server->exec()\\n#20 \\\/srv\\\/www\\\/owncloud8.1-locking\\\/remote.php(84): require_once('\\\/srv\\\/www\\\/ownclo...')\\n#21 {main}\",\"File\":\"\\\/srv\\\/www\\\/owncloud8.1-locking\\\/apps\\\/encryption\\\/lib\\\/crypto\\\/encryption.php\",\"Line\":286}","level":4,"time":"2015-06-19T12:51:52+00:00"}
{"reqId":"vBp3myXQ0G5BvncRJ67S","remoteAddr":"148.251.209.108","app":"PHP","message":"Call to a member function getPath() on boolean at \/srv\/www\/owncloud8.1-locking\/lib\/private\/share\/share.php#132","level":3,"time":"2015-06-19T12:52:01+00:00"}

**Comment by DeepDiver1975** [comment]
a lock'n'crypt error?

**Comment by PVince81** [comment]
Confirmed, lock'n'crypt:
```
2015-06-19 16:14:24,675 - ERROR - recipient_worker - Unexpected response: Status code: 500 - 
2015-06-19 16:14:24,696 - ERROR - recipient_worker - Operation "move_in" must be allowed when share permissions are 31 'Operation "%s" must be %s when share permissions are %i' % (operation, success_message, perms) failed in recipient_worker() ["/home/vincent/work/workspace/smashbox/lib/oc-tests/test_sharePermissions.py" at line 251]
INFO:requests.packages.urllib3.connectionpool:Resetting dropped connection: 127.0.0.1
2015-06-19 16:14:26,340 - ERROR - recipient_worker - Unexpected response: Status code: 500 - 
2015-06-19 16:14:26,341 - ERROR - recipient_worker - Operation "move_in_overwrite" must be allowed when share permissions are 31 'Operation "%s" must be %s when share permissions are %i' % (operation, success_message, perms) failed in recipient_worker() ["/home/vincent/work/workspace/smashbox/lib/oc-tests/test_sharePermissions.py" at line 251]
INFO:requests.packages.urllib3.connectionpool:Resetting dropped connection: 127.0.0.1
2015-06-19 16:14:29,001 - ERROR - recipient_worker - Unexpected response: Status code: 500 - 
2015-06-19 16:14:29,002 - ERROR - recipient_worker - Operation "move_out" must be allowed when share permissions are 31 'Operation "%s" must be %s when share permissions are %i' % (operation, success_message, perms) failed in recipient_worker() ["/home/vincent/work/workspace/smashbox/lib/oc-tests/test_sharePermissions.py" at line 251]
INFO:requests.packages.urllib3.connectionpool:Resetting dropped connection: 127.0.0.1
2015-06-19 16:14:33,294 - ERROR - recipient_worker - 3 error(s) reported
CRITICAL:smash.:Aborting run -- non-zero exit code (2)
```

@icewind1991 please try again.

**Comment by PVince81** [comment]
Very strange, on the same instance I did this:
1. Create folder "test"
2. Upload picture inside (web UI)

getting this:
```
{"reqId":"ZyGTAwkpnNh91UIRMSZ9","remoteAddr":"127.0.0.1","app":"PHP","message":"Call to a member function getPath() on boolean at \/srv\/www\/htdocs\/owncloud\/lib\/private\/share\/share.php#132","level":3,"time":"2015-06-19T14:20:03+00:00","method":"POST","url":"\/owncloud\/index.php\/apps\/files\/ajax\/upload.php"}
```

**Comment by DeepDiver1975** [comment]
Sounds like Friday afternoon ... time to shutdown and head off to the weekend.

**Comment by PVince81** [comment]
@icewind1991 have you been able to reproduce this ?

**Comment by nickvergessen** [comment]
The problem is https://github.com/owncloud/core/commit/72eedda16cec0b15850371035c517e98b0e9e064?w=1#diff-d9747e68e80ce288dcea911cb8607546R1179

because the file is new, (just being uploaded) there is no old cache data?The question is, why are we receiving a locked exception anyway, because we have the lock that created the file?

**Comment by nickvergessen** [comment]
Summary of the problem:
1. We create a file
2. We lock it to write the content (View::file_put_content: https://github.com/owncloud/core/blob/master/lib/private/files/view.php#L550-550 )
3. The encryption wrapper gets a list of users on fopen: https://github.com/owncloud/core/blob/master/lib/private/files/view.php#L554 => https://github.com/owncloud/core/blob/master/lib/private/encryption/file.php#L57-57
4. The share list gets the file info: https://github.com/owncloud/core/blob/master/lib/private/share/share.php#L131-131 which tries to get a lock itself: https://github.com/owncloud/core/blob/master/lib/private/files/view.php#L1164

So we try to have a shared and exclusive lock for the same file inside the same php request. Should we try to bind locks to request ids to avoid stuff like this?

@icewind1991

**Comment by icewind1991** [comment]
> Should we try to bind locks to request ids to avoid stuff like this?

I dont think so, even within the same request locks should be abided.

I think the problem comes down to
```php
if ($view->file_exists($path)) {
	$meta = $view->getFileInfo($path);
	$path = substr($meta->getPath(), strlen('/' . $ownerUser . '/files'));
} else {
....
}
```

the call to `fopen` creates the file and `file_exists` "sees" the new file since `file_exists` doesn't lock while `getFileInfo` doesn't since that does lock.

Maybe it would be better to have something like
```php
$meta = $view->getFileInfo($path);
if ($meta) {
	$path = substr($meta->getPath(), strlen('/' . $ownerUser . '/files'));
} else {
...
}

**Comment by PVince81** [comment]
Ok, I got it happening again with the smashbox tests.
And also with the following steps:
1. "user1" shares folder "folder" with "user2"
2. Login as "user2" with cadaver
3. Put a file "test.txt" in the root
4. Move "test.txt" into "folder/test.txt"

Getting:
```
{"reqId":"TEl1aJwdi9idLlBhms9M","remoteAddr":"127.0.0.1","app":"PHP","message":"Call to a member function getPath() on boolean at \/srv\/www\/htdocs\/owncloud\/lib\/private\/share\/share.php#132","level":3,"time":"2015-06-24T15:46:05+00:00","method":"MOVE","url":"\/owncloud\/remote.php\/webdav\/test.txt"}
```

**Comment by PVince81** [comment]
The stack trace leading to the issue:
```
0  OC\Share\Share::getUsersSharingFile() /srv/www/htdocs/owncloud/lib/private/share/share.php:132
1  OCP\Share::getUsersSharingFile() /srv/www/htdocs/owncloud/lib/public/share.php:88
2  OC\Encryption\File->getAccessList() /srv/www/htdocs/owncloud/lib/private/encryption/file.php:57
3  OC\Files\Stream\Encryption->stream_open() /srv/www/htdocs/owncloud/lib/private/files/stream/encryption.php:246
4  fopen()         /srv/www/htdocs/owncloud/lib/private/files/stream/encryption.php:187
5  OC\Files\Stream\Encryption::wrapSource() /srv/www/htdocs/owncloud/lib/private/files/stream/encryption.php:187
6  OC\Files\Stream\Encryption::wrap() /srv/www/htdocs/owncloud/lib/private/files/stream/encryption.php:169
7  OC\Files\Storage\Wrapper\Encryption->fopen() /srv/www/htdocs/owncloud/lib/private/files/storage/wrapper/encryption.php:412
8  OC\Files\Storage\Wrapper\Encryption->copyBetweenStorage() /srv/www/htdocs/owncloud/lib/private/files/storage/wrapper/encryption.php:502
9  OC\Files\Storage\Wrapper\Encryption->moveFromStorage() /srv/www/htdocs/owncloud/lib/private/files/storage/wrapper/encryption.php:436
10 OC\Files\Storage\Wrapper\Wrapper->moveFromStorage() /srv/www/htdocs/owncloud/lib/private/files/storage/wrapper/wrapper.php:535
11 OC\Files\Storage\Shared->moveFromStorage() /srv/www/htdocs/owncloud/apps/files_sharing/lib/sharedstorage.php:612
12 OC\Files\Storage\Wrapper\Wrapper->moveFromStorage() /srv/www/htdocs/owncloud/lib/private/files/storage/wrapper/wrapper.php:535
13 OC\Files\View->rename() /srv/www/htdocs/owncloud/lib/private/files/view.php:680
14 OC\Connector\Sabre\ObjectTree->move() /srv/www/htdocs/owncloud/lib/private/connector/sabre/objecttree.php:232
15 Sabre\DAV\CorePlugin->httpMove() /srv/www/htdocs/owncloud/3rdparty/sabre/dav/lib/DAV/CorePlugin.php:630
16 call_user_func_array:{/srv/www/htdocs/owncloud/3rdparty/sabre/event/lib/EventEmitterTrait.php:105}() /srv/www/htdocs/owncloud/3rdparty/sabre/event/lib/EventEmitterTrait.php:105
17 Sabre\Event\EventEmitter->emit() /srv/www/htdocs/owncloud/3rdparty/sabre/event/lib/EventEmitterTrait.php:105
18 Sabre\DAV\Server->invokeMethod() /srv/www/htdocs/owncloud/3rdparty/sabre/dav/lib/DAV/Server.php:469
19 Sabre\DAV\Server->exec() /srv/www/htdocs/owncloud/3rdparty/sabre/dav/lib/DAV/Server.php:254
20 require_once()  /srv/www/htdocs/owncloud/apps/files/appinfo/remote.php:83
21 {main}          /srv/www/htdocs/owncloud/remote.php:84
```

**Comment by PVince81** [comment]
Huh, woot?

So, with encryption a cross-storage move is basically a copy.
It does fopen('test.txt', 'r') and fopen('folder/test.txt', 'w')

The read call calls `getAccessList('/user2/files/test.txt')` 
and the write call: `getAccessList('/user1/files/folder/test.txt'`).  Obviously, this file does not exist, so `getFileInfo()` will fail!

**Comment by PVince81** [comment]
When locking is disabled, the target file already exists, so the second getFileInfo works.

**Comment by PVince81** [comment]
Ok yes, I understand, as stated by @nickvergessen here https://github.com/owncloud/core/issues/17029#issuecomment-114050620

If we could do getFileInfo before fopen (and cache the result?) then this wouldn't be an issue for this specific code path.

Now the question whether getAccessList() can be done outside of stream_open, before ? @schiesbn what do you think ?

**Comment by PVince81** [comment]
In the stack trace I see that the fopen() is done in wrapSource: https://github.com/owncloud/core/blob/master/lib/private/files/stream/encryption.php#L187

Maybe `getAccessList` could be called there and stored into the stream context ?

**Comment by schiesbn** [comment]
> So, with encryption a cross-storage move is basically a copy.

not only with encryption, see https://github.com/owncloud/core/blob/master/lib/private/files/storage/common.php#L585

**Comment by schiesbn** [comment]
> The read call calls getAccessList('/user2/files/test.txt')
> and the write call: getAccessList('/user1/files/folder/test.txt'). Obviously, this file does not exist, so getFileInfo() will fail!

We check if the file exists and call getAccessList for the parent folder in this case: https://github.com/owncloud/core/blob/master/lib/private/files/stream/encryption.php#L242

**Comment by schiesbn** [comment]
If I understand the explanation of @nickvergessen correctly we already obtain the lock in the view on file_put_contents. If this is the case I don't see a difference in calling getAccessList() before or after we wrap the stream. The first lock will always be already there.

**Comment by icewind1991** [comment]
imo the way to solve this would to have `getAccessList` work correctly on files that aren't saved in the cache yet (the file does exist since fopen($path,'w') creates it but the cache is only updated after the write is complete) #17071 was an attempt at that.

**Comment by nickvergessen** [comment]
From @icewind1991 at https://github.com/owncloud/core/pull/17164#issuecomment-115215965
> I don't think this is the correct way to solve the problem, having the lock actually lock the file even for the current process prevents getting unexpected results when trying to do fs work in the middle of an existing fs operation, having a storage wrapper call `getFileInfo` on a file that hasn't been created yet should not return an unspecified maybe-valid-but-relies-completely-on-the-storage-backend result, this should return an error value (or an exception, a man can dream...).
>
> A write lock is not just to keep other processes from reading the file, it should also prevent the same process from reading the file since we can't guarantee correct behavior when reading a file that's being written to.
>
> In the past we've had to fix multiple bugs that we're caused by wrappers/hooks/etc interfering with filesystem operations. Locking should make filesystem operations atomic and prevent unexpected behavior, something we cant do if the same process can simply ignore the lock.

**Comment by nickvergessen** [comment]
> Locking should make filesystem operations atomic and prevent unexpected behavior, something we cant do if the same process can simply ignore the lock.

Well I see the idea behind that, the problem is... they are not. Any hook, storage wrapper, ... that is called in the action, needs to be able to behave like it did in the past. This is exactly why our test is failing atm. Getting the user's an item is shared with is not something randoom (typo intended).
It is done by encryption (to generate/copy keys), activity (to generate the activities), ...

The patch you suggest at https://github.com/owncloud/core/pull/17071 falls back to the parent item. However, that information is simply wrong. If a file is shared directly, the directory will not contain those users, which is why the smashbox execution I did fails (it fails differently compared to before the patch):

> OC\Encryption\Exceptions\DecryptionFailedException - Message: Can not decrypt this file, probably this is a shared file. Please ask the file owner to reshare the file with you.

So the keys of the people sharing the file but not any parent, are not there anymore (assumption, but somehow some keys are missing). We could try and gather all necessary information, before creating the lock. But this means huge overhead (we don't know which methods will be called, so we have to call all of them ....) and then we would still need to tunnel all information to all methods out there, which currently call any filesystem related method directly.

Feel free to suggest a new way, that is "*better*" then #17164 while still solving the issues we are seeing. At least #17071 is not the right solution.

**Comment by PVince81** [comment]
This issue also seems to break regular file upload, see https://github.com/owncloud/core/issues/17153

**Comment by PVince81** [comment]
but smashbox didn't find it because for some mysterious reasons WebDAV upload works and doesn't reach that code path. However web UI upload triggers the issue.

Will investigate WebDAV separately.

**Comment by PVince81** [comment]
Ok got it: when uploading small files with WebDAV (non-chunking), we only set a shared lock on the target file https://github.com/owncloud/core/blob/master/lib/private/connector/sabre/file.php#L117. Then we write into a PART file, so at this point we can still call getFileInfo on the final file path. It's only during the final rename https://github.com/owncloud/core/blob/master/lib/private/connector/sabre/file.php#L205  that the target path gets an exclusive lock.

So basically if we used PART files for cross-storage moves, we could use the same logic eventually... (#13379)

**Comment by icewind1991** [comment]
Note that is only used to pass by chance, when using a local storage backend the `fopen` call creates the file, with most external storages the file will only be created after the file handle is closed so it has always been broken on external storages

**Comment by nickvergessen** [comment]
So @icewind1991 did you have a better idea? I know locking is experimental, but not being able to do webdav move_in, move_in_overwrite and move_out seems like a real draw back, just for the sane of theory-crafting which can not be met at the current time.

If not then I will fix the unit tests on my PR so they pass and we can go with the non-ideal-world solution rather than a broken one, deal?

cc @DeepDiver1975

**Comment by PVince81** [comment]
This is worse than just move in: web upload is also broken.

**Comment by icewind1991** [comment]
> web upload is also broken.

Ok I was able to reproduce that, will look into it

**Comment by nickvergessen** [comment]
Seems like #17030 and #17071 had sufficient success. Tests are passing again: closing.
